testsuite/tests/typing-modules/include_functor.ml

```diff
@@ -430,8 +430,192 @@ module F14 : functor (X : S) (Y : S) -> sig val z : X.t * Y.t end
+9 |   include functor F14_2
+                      ^^^^^
+Error: The type of this functor's result is not includable; it is
+       functor () () -> sig val z : X.t end
+|}];;
+
+(* Test 15: Make sure we're extracting functor return types appropriately *)
+module type S15 = sig val x : int end
+module type S15' = S15
+
+module F15 (X : sig end) : S15' =
+struct
+  let x = 42
+end
+
+include functor F15
+
+[%%expect{|
+module type S15 = sig val x : int end
+module type S15' = S15
+Line 4, characters 12-13:
+4 | module F15 (X : sig end) : S15' =
+                ^
+Warning 60 [unused-module]: unused module X.
+module F15 : functor (X : sig end) -> S15'
+val x : int = 42
+|}]
+
+(* Test 16: Make sure we're adequately finding the definitions of module type
+   identifiers *)
+module type S16 = sig
+  type t
+  val x : t
+end
+
+module type S16' = sig
+  type s
+end
+
+module type F16_1 = functor (X : S16) -> S16'
+module type F16_2 = functor (_ : S16) -> S16'
+
+module type G16_1 = sig
+  type t
+  val x : t
+
+  include functor F16_1
+end
+
+module type G16_2 = sig
+  type t
+  val x : t
+
+  include functor F16_2
+end;;
+[%%expect{|
+module type S16 = sig type t val x : t end
+module type S16' = sig type s end
+Line 10, characters 29-30:
+10 | module type F16_1 = functor (X : S16) -> S16'
+                                  ^
+Warning 67 [unused-functor-parameter]: unused functor parameter X.
+module type F16_1 = functor (X : S16) -> S16'
+module type F16_2 = S16 -> S16'
+module type G16_1 = sig type t val x : t type s end
+module type G16_2 = sig type t val x : t type s end
+|}];;
+
+(* Test 17: Functors whose types don't begin with a normal applicative parameter
+   are rejected. *)
+module type S17 = sig
+  type t
+  val x : t
+end
+
+module type F17_1 = functor () -> S17
+
+module type G17 = sig
+  include functor F17_1
+end;;
+[%%expect {|
+module type S17 = sig type t val x : t end
+module type F17_1 = functor () -> S17
+Line 9, characters 18-23:
+9 |   include functor F17_1
+                      ^^^^^
+Error: The type of this functor is: functor () -> S17.
+       Its parameter is not a signature.
+|}];;
+
+module type F17_2 = functor () (X : S17) -> sig val z : X.t end
+
+module type G17_2 = sig
+  type t
+  val x : t
+  include functor F17_2
+end;;
+[%%expect {|
+module type F17_2 = functor () (X : S17) -> sig val z : X.t end
+Line 6, characters 18-23:
+6 |   include functor F17_2
+                      ^^^^^
+Error: The type of this functor is:
+       functor () (X : S17) -> sig val z : X.t end.
+       Its parameter is not a signature.
+|}];;
+
+(* Test 18: Generative functors *)
+module type S18 = sig
+  type t
+  val x : t
+  val equal : t -> t -> bool
+end
+
+module F18 (X : S18) () : sig
+  type t'
+  val z : t'
+  val equal_t' : t' -> t' -> bool
+end = struct
+  type t' = X.t
+  let z = X.x
+  let equal_t' = X.equal
+end
+
+module M18 = struct
+  type t = int
+  let x = 42
+  let equal = Int.equal
+  include functor F18
+end
+
```

Comment:
Maybe add a second test showing that you can't include a generative functor inside an applicative functor